### ISSUE TITLE
dagpool_test busy loop on CPython 3.7 epool/select hubs

### ISSUE THREAD

**Issue by temoto** [eventlet/eventlet#475]
While executing dagpool_test, random tests will burn CPU in kernel. Observed as

```
./venv-37/bin/nosetests -svx tests/dagpool_test.py
tests.dagpool_test.test_spawn_collision_spawn ... ok
tests.dagpool_test.test_spawn_multiple ... ok
tests.dagpool_test.test_spawn_many ... ok
tests.dagpool_test.test_wait_each_all ... ok
tests.dagpool_test.test_kill ... ok
tests.dagpool_test.test_post_collision_preload ... ok
tests.dagpool_test.test_post_collision_post ... ok
tests.dagpool_test.test_post_collision_spawn ... ok
tests.dagpool_test.test_post_replace ... ok
tests.dagpool_test.test_getitem ... ok
tests.dagpool_test.test_waitall_exc ... ok
tests.dagpool_test.test_propagate_exc ... ^C^C^\[1]    26649 quit       ./venv-37/bin/nosetests -svx tests/dagpool_test.py
   23.65s user 0.06s system 99% cpu 23.727 total
```
Note system 99%.

- Nose process is not responding to SIGINT (KeyboardInterrupt).
- Particular tests to cause busy loop: test_propagate_exc, test_wait_each_exc.
- They pass alone, problem occurs while running whole dagpool_test.
- Environment: Linux, python 3.7, epoll and select hubs

Example of build failed for this reason:
https://travis-ci.org/eventlet/eventlet/jobs/351626233

@nat-goodspeed could you look what is the problem here?

Similar busy loop problem with `test_accept_deflate_ext_context_takeover_13 (tests.websocket_new_test.TestWebSocketWithCompression)`. Difference is: under strace websocket tests segfault with 

```
recvfrom(9, "", 2, 0, NULL, NULL)       = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x99} ---
```

and dagpool doesn't

```
write(2, "tests.dagpool_test.test_propagate_exc ... ", 42tests.dagpool_test.test_propagate_exc ... ) = 42
clock_gettime(CLOCK_MONOTONIC, {3917231, 687934179}) = 0
clock_gettime(CLOCK_MONOTONIC, {3917231, 687991532}) = 0
clock_gettime(CLOCK_MONOTONIC, {3917231, 688159907}) = 0
clock_gettime(CLOCK_MONOTONIC, {3917231, 688235242}) = 0
clock_gettime(CLOCK_MONOTONIC, {3917231, 688261079}) = 0
^C--- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
```

**Comment by temoto** [comment]
Current status: 2018-03-10 on Linux CPython 3.7 Eventlet may cause busy loop, use in production is not recommended, subscribe to this issue to get updates.

**Comment by nat-goodspeed** [comment]
Sorry, I'm at a conference all this week. I will look into it next week.

**Comment by nat-goodspeed** [comment]
Are you building Python 3.7 from source? I tried to find it on python.org, but it seems to be unavailable as yet. That makes it tough to debug: I don't know how to obtain the interpreter version with which this fails.

If Python 3.7 has not yet been released, isn't it possible that what you're seeing is a bug in Python 3.7?

**Comment by temoto** [comment]
@nat-goodspeed it is there, https://www.python.org/downloads/release/python-370b3/

3.7-dev supplied by Travis also produces this problem.

**Comment by nat-goodspeed** [comment]
Ah, so Python 3.7 was officially released yesterday. There goes my hope that it was a bug in the pre-release version. I haven't forgotten about this, really!

**Comment by vstinner** [comment]
I created an upstream bug report: https://bugs.python.org/issue33996

I'm able to reproduce the crash without eventlet, only using greenlet.

**Comment by vstinner** [comment]
>  Doesn't that mean the fix for Python 3.7 would be in greenlet, though, rather than the Python interpreter?

I wrote a script to reproduce the bug only using greenlet. Maybe it's a bug in greenlet, maybe in Python. Honestly, I don't know at this point. At least, it doesn't seem to be a bug in eventlet.

**Comment by vstinner** [comment]
The bug is in greenlet, not in Python: https://github.com/python-greenlet/greenlet/issues/131

**Comment by hroncok** [comment]
With @vstinner'spatch, this issue is solved, yet eventlet still fails with 3.7:

```
======================================================================
ERROR: test_connect_ssl (tests.api_test.TestApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/api_test.py", line 67, in test_connect_ssl
    server_side=True
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_wrap_ssl (tests.convenience_test.TestServe)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/convenience_test.py", line 121, in test_wrap_ssl
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_wrap_socket (tests.greenio_test.TestGreenSocket)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/greenio_test.py", line 476, in test_wrap_socket
    ssl.wrap_socket(sock)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_duplex_response (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 45, in test_duplex_response
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_greensslobject (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 192, in test_greensslobject
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_no_handshake_block_accept_loop (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 210, in test_no_handshake_block_accept_loop
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 278, in test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().
-------------------- >> begin captured stdout << ---------------------
Trying <function SSLTest.test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data.<locals>.recv at 0x7f72781dea60>...

--------------------- >> end captured stdout << ----------------------

======================================================================
ERROR: test_recv_after_ssl_connect (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 93, in test_recv_after_ssl_connect
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_regression_gh_17 (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 202, in test_regression_gh_17
    sock = ssl.wrap_socket(socket.socket())
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: SSL socket.sendall() busy loop
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 173, in test_sendall_cpu_usage
    server_sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_close (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 63, in test_ssl_close
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_connect (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 78, in test_ssl_connect
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_unwrap (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 140, in test_ssl_unwrap
    client_ssl = ssl.wrap_socket(client)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_sending_messages (tests.websocket_test.TestWebSocketSSL)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/websocket_test.py", line 541, in test_ssl_sending_messages
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_012_ssl_server (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 451, in test_012_ssl_server
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_013_empty_return (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 473, in test_013_empty_return
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_017_ssl_zeroreturnerror (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 578, in test_017_ssl_zeroreturnerror
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_028_ssl_handshake_errors (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 1110, in test_028_ssl_handshake_errors
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
FAIL: test_line (tests.debug_test.TestSpew)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/debug_test.py", line 39, in test_line
    assert "f=<frame object at" in output
AssertionError

======================================================================
FAIL: test_line_global (tests.debug_test.TestSpew)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/debug_test.py", line 62, in test_line_global
    assert "f=<frame object at" in output
AssertionError

----------------------------------------------------------------------
Ran 671 tests in 49.917s

FAILED (SKIP=95, errors=18, failures=2)

```

**Comment by vstinner** [comment]
> With @vstinner'spatch, this issue is solved, yet eventlet still fails with 3.7: (...)

I suggest to open a new issue to track these issues, since they are different (unrelated) ;-)

**Comment by nat-goodspeed** [comment]
None of the test failures in #502 seem related to dagpool_test. Closing.
